addon/components/sidenotes-wrapper.js

```diff
@@ -0,0 +1,89 @@
+import Component from '@glimmer/component';
+import { action } from '@ember/object';
+import { tracked } from '@glimmer/tracking';
+import { debounce } from '@ember/runloop';
+
+import { getNotesIdealPlacement } from '../utils/notes-placement';
+
+export default class SidenotesWrapperComponent extends Component {
+  @tracked _selectedSidenoteId = null;
+  @tracked _alignedSidenoteId = null;
+
+  @tracked notes = [];
+
+  get gutter() {
+    return this.args.gutter ?? 0;
+  }
+
+  get controlledMode() {
+    return this.args.selectedSidenoteId !== undefined;
+  }
+
+  constructor() {
+    super(...arguments);
+
+    this._selectedSidenoteId = this._alignedSidenoteId = this.controlledMode
+      ? this.args.selectedSidenoteId
+      : this.args.defaultSidenoteId ?? null;
+  }
+
+  placeNotes() {
+    const idealPlacement = getNotesIdealPlacement(
+      this.notes,
+      this._alignedSidenoteId,
+      this.gutter
+    );
+
+    idealPlacement.forEach(({ top, note: { element } }) => {
+      element.style.top = `${top}px`;
+    });
+
+    this.args.onSidenotesMoved?.();
+  }
+
+  @action
+  onRegisterSidenote({ element, id, offsetY, data }) {
+    this.notes.push({ element, id, offsetY, data });
+
+    if (this.notes.length === this.args.items.length) {
+      debounce(this, this.placeNotes, 0);
+    }
+  }
+
+  @action
+  onSelectedSidenoteIdUpdate() {
+    const { selectedSidenoteId } = this.args;
+
+    this._selectedSidenoteId = selectedSidenoteId;
+    if (selectedSidenoteId) {
+      this._alignedSidenoteId = selectedSidenoteId;
+    }
+    debounce(this, this.placeNotes, 0);
```

Comment:
If you always call `placeNotes` with debounce you should place it inside the function or at least have an intermediate function to factorize it

Reply:
intermediate function created https://github.com/concordnow/ember-sidenotes/pull/1/commits/e1638dafae73ae8cd3c2330f0d54f66511229d1b :+1: